### PR TITLE
Testing stats

### PR DESCRIPTION
**Pull request by arussellsaw**
i added asserts and test counts, if you know how to get ast.CallExpr.Fun as a string i could make it a bit better, but i didn't try very hard.

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -63,6 +67,9 @@ func (p *Parser) ParseDir(targetDir string) *Result {
 					res.Function++
 					if x.Name.IsExported() {
 						res.ExportedFunction++
+						if strings.HasPrefix(x.Name.String(), "Test") {
+							res.Tests++
```
Comment (warmans) [inline main.go]: I think this needs to check for the testing.t function argument or it could have false positives.

Comment (arussellsaw) [inline main.go]: i thought Test was a reserved prefix, but you're right, i'll add that

File: main.go [hunk 2]
```diff
@@ -133,20 +154,62 @@ func (p *Parser) CountLOC(filePath string) (int, int) {
...
 }
 
-func (t *TextReport) Print(res *Result) {
-	fmt.Printf("LOC:        %v (%v CLOC, %v NCLOC)\n", res.LOC, res.CLOC, res.NCLOC)
-	fmt.Printf("Imports:    %v\n", res.Import)
-	fmt.Printf("Structs:    %v\n", res.Struct)
-	fmt.Printf("Interfaces: %v\n", res.Interface)
-	fmt.Printf("Methods:    %v (%v Exported)\n", res.Method, res.ExportedMethod)
-	fmt.Printf("Functions:  %v (%v Exported)\n", res.Function, res.ExportedFunction)
+func (t *TextReport) Print(res *Result, outputFmt string) {
```
Comment (warmans) [inline main.go]: This is slightly wrong. There should be an interface called ReportInterface which just has the print method. Then each report format (TextReport, JsonReport, etc.) can be implemented as a struct implementing that interface. You still need a switch in the main() function to create the required report format.

File: main.go [hunk 3]
```diff
@@ -24,10 +26,15 @@ type Result struct {
 	Function         int
 	ExportedFunction int
 	Import           int
+	Tests            int
+	TLOC             int
```
Comment (warmans) [inline main.go]: This property is unused.

File: main.go [hunk 4]
```diff
@@ -63,6 +69,23 @@ func (p *Parser) ParseDir(targetDir string) *Result {
...
+						if strings.HasPrefix(x.Name.String(), "Test") {
+							nodePos := fset.Position(x.Type.Params.List[0].Type.Pos())
+							nodeEnd := fset.Position(x.Type.Params.List[0].Type.End())
+							nodeFile, _ := os.Open(nodePos.Filename)
+							defer nodeFile.Close()
+							node := make([]byte, (nodeEnd.Offset - nodePos.Offset))
+							nodeFile.ReadAt(node, int64(nodePos.Offset))
+							paramTypes := []string{
+								"*testing.T",
+								"*testing.M",
```
Comment (warmans) [inline main.go]: there is testing.B as well. Might be worth checking for.

**Comment by arussellsaw** [comment]
should be okay now...

**Comment by warmans** [comment]
Are you going to add testing.B to the tests count?

**Comment by arussellsaw** [comment]
didn't check the testing package for all types, just went on memory, obviously a mistake.

**Comment by warmans** [comment]
Oh forgot to say, could you add some tests for the new counts?

**Pull request merged**